### ISSUE TITLE
Log sometimes contains System.InvalidOperationException: Collection was modified

### ISSUE THREAD

**Issue by TraGicCode** [Particular/ServiceControl#1550]
Hey Guys,

I currently have the latest version of ServiceControl 3.5.0.  I noticed that sometimes the log contains the following information about an exception that is occuring. 

```
2018-12-28 13:23:23.8650|268|Info|ServiceControl.Recoverability.RetryProcessor|No batch found to forward.
2018-12-28 13:23:23.8650|268|Info|ServiceControl.Recoverability.RetryProcessor|No batch found to stage.
2018-12-28 13:23:27.8650|123|Info|NServiceBus.RecoverabilityExecutor|Immediate Retry is going to retry message '4165fbb8-99a0-4bad-a0fb-f67e261de6f2' because of an exception:
System.InvalidOperationException: Collection was modified; enumeration operation may not execute.
   at System.ThrowHelper.ThrowInvalidOperationException(ExceptionResource resource)
   at System.Collections.Generic.Dictionary`2.Enumerator.MoveNext()
   at NServiceBus.Extensibility.ContextBag.Merge(ContextBag context)
   at NServiceBus.MainPipelineExecutor.<Invoke>d__1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NServiceBus.Transport.RabbitMQ.MessagePump.<Process>d__29.MoveNext()
```

This sounds like a bug In ServiceControl somewhere.  While it's not causing servicecontrol to crash and the error goes doesn't occur further in the logs i figured i would report this.

**Comment by mikeminutillo** [comment]
@TraGicCode how often does this exception come up? Can you send some log files through to support? If you can include the diagnostic file that is dropped into the log folder that would be great as well.

I'll see if I can reproduce the issue locally.

**Comment by TraGicCode** [comment]
Hey @mikeminutillo ,

I'll get back to you with more details towards the end of this up coming week.

**Comment by TraGicCode** [comment]
I will post here every time i notice it happen with detailed diag information.

**Comment by TraGicCode** [comment]
Occurrence 2:

ServiceControl Instance's log file
```
2019-01-06 19:04:49.9760|34|Info|ServiceControl.Recoverability.RetryProcessor|No batch found to stage.
2019-01-06 19:05:05.3823|23|Info|NServiceBus.RecoverabilityExecutor|Immediate Retry is going to retry message 'cdc3ec8d-e333-41ef-8640-a9ce0011e082' because of an exception:
System.InvalidOperationException: Collection was modified; enumeration operation may not execute.
   at System.ThrowHelper.ThrowInvalidOperationException(ExceptionResource resource)
   at System.Collections.Generic.Dictionary`2.Enumerator.MoveNext()
   at NServiceBus.Extensibility.ContextBag.Merge(ContextBag context)
   at NServiceBus.MainPipelineExecutor.<Invoke>d__1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NServiceBus.Transport.RabbitMQ.MessagePump.<Process>d__29.MoveNext()
2019-01-06 19:05:19.9917|30|Info|ServiceControl.Recoverability.RetryProcessor|No batch found to forward.
2019-01-06 19:05:19.9917|30|Info|ServiceControl.Recoverability.RetryProcessor|No batch found to stage.
```

**Comment by SzymonPobiega** [comment]
@TraGicCode thanks for reporting it. We have found a concurrency problem in the RabbitMQ transport and we will be fixing it shortly.

**Comment by danielmarbach** [comment]
@TraGicCode we are about to fix a release for this. It will be fixed as part of 3.6.1

**Comment by TraGicCode** [comment]
Thanks @danielmarbach
